benchmark_decode_test.go

```diff
@@ -0,0 +1,312 @@
+package redis_test
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-redis/redis/v8"
+	"github.com/go-redis/redis/v8/internal/proto"
+	"net"
+	"strings"
+	"testing"
+	"time"
+)
+
+type Conn struct {
+	buff []byte
+}
+
+func (c *Conn) Read(b []byte) (n int, err error)   { return copy(b, c.buff), nil }
+func (c *Conn) Write(b []byte) (n int, err error)  { return len(b), nil }
+func (c *Conn) Close() error                       { return nil }
+func (c *Conn) LocalAddr() net.Addr                { return nil }
+func (c *Conn) RemoteAddr() net.Addr               { return nil }
+func (c *Conn) SetDeadline(_ time.Time) error      { return nil }
+func (c *Conn) SetReadDeadline(_ time.Time) error  { return nil }
+func (c *Conn) SetWriteDeadline(_ time.Time) error { return nil }
+
+func NewDecodeClient() *RespDecode {
+	conn := &Conn{}
+	client := redis.NewClient(&redis.Options{
+		PoolSize: 128,
+		Dialer: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return conn, nil
+		},
+	})
+	return &RespDecode{
+		name:   "Client",
+		conn:   conn,
+		client: client,
+	}
+}
+
+func NewDecodeClusterClient() *RespDecode {
+	conn := &Conn{}
+	client := redis.NewClusterClient(&redis.ClusterOptions{
+		PoolSize: 128,
+		Addrs:    []string{"127.0.0.1:6379"},
+		Dialer: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return conn, nil
+		},
+		ClusterSlots: func(_ context.Context) ([]redis.ClusterSlot, error) {
+			return []redis.ClusterSlot{
+				{
+					Start: 0,
+					End:   16383,
+					Nodes: []redis.ClusterNode{{Addr: "127.0.0.1:6379"}},
+				},
+			}, nil
+		},
+	})
+	// init command.
+	cmd := []string{
+		// set
+		"*7\r\n+set\r\n:-3\r\n*2\r\n+write\r\n+denyoom\r\n:1\r\n:1\r\n:1\r\n" +
+			"*3\r\n+@write\r\n+@string\r\n+@slow\r\n",
+
+		// get
+		"*7\r\n+get\r\n:2\r\n*2\r\n+readonly\r\n+fast\r\n:1\r\n:1\r\n:1\r\n" +
+			"*3\r\n+@read\r\n+@string\r\n+@fast\r\n",
+
+		// del
+		"*7\r\n+del\r\n:-2\r\n*1\r\n+write\r\n:1\r\n:-1\r\n:1\r\n" +
+			"*3\r\n+@keyspace\r\n+@write\r\n+@slow\r\n",
+
+		// incr
+		"*7\r\n+incr\r\n:2\r\n*3\r\n+write\r\n+denyoom\r\n+fast\r\n:1\r\n:1\r\n:1\r\n" +
+			"*3\r\n+@write\r\n+@string\r\n+@fast\r\n",
+
+		// mget
+		"*7\r\n+mget\r\n:-2\r\n*2\r\n+readonly\r\n+fast\r\n:1\r\n:-1\r\n:1\r\n" +
+			"*3\r\n+@read\r\n+@string\r\n+@fast\r\n",
+
+		// multi
+		"*7\r\n+multi\r\n:1\r\n*4\r\n+noscript\r\n+loading\r\n+stale\r\n+fast\r\n:0\r\n:0\r\n:0\r\n" +
+			"*2\r\n+@fast\r\n+@transaction\r\n",
+
+		// exec
+		"*7\r\n+exec\r\n:1\r\n*5\r\n+noscript\r\n+loading\r\n+stale\r\n+skip_monitor\r\n+skip_slowlog\r\n" +
+			":0\r\n:0\r\n:0\r\n*2\r\n+@slow\r\n+@transaction\r\n",
+
+		// ping
+		"*7\r\n+ping\r\n:-1\r\n*2\r\n+stale\r\n+fast\r\n:0\r\n:0\r\n:0\r\n*2\r\n+@fast\r\n+@connection\r\n",
```

Comment:
I don't understand the purpose of these commands... As I understand `conn.buff` is a response buffer, but these are outgoing commands...

Reply:
The cluster uses the `command` command (cmdsInfoCache), but does not need to respond to all commands (200+), only the commands used in our test

Reply:
Did you consider setting `client.cmdsInfoCache` instead? We could even grab `client.cmdsInfoCache` from the real cluster client so we don't need to construct it ourselves.

Reply:
I modified to get cmdsInfoCache from the real redis-server, what I originally thought was not to rely on redis-server

Reply:
Thanks, it looks much more approachable now.